File: 2025/day04/py/day04.py
from collections import defaultdict


Grid = defaultdict
grid_row_cnt = 0
grid_col_cnt = 0


def parse(filename) -> Grid:
    global grid_row_cnt, grid_col_cnt

    with open(filename) as f:
        lines = [line.strip() for line in f.readlines()]

    grid_row_cnt = len(lines)
    grid_col_cnt = len(lines[0])

    grid = Grid(lambda: ".")

    for r in range(grid_row_cnt):
        for c in range(grid_col_cnt):
            grid[(r, c)] = lines[r][c]

    return grid


def find_removable(grid: Grid) -> list[tuple[int, int]]:
    global grid_row_cnt, grid_col_cnt

    removable = []
    for r in range(grid_row_cnt):
        for c in range(grid_col_cnt):
            curr_val = grid[(r, c)]
            if curr_val == "@":
                # get neighbors in all 8 directions
                neighbor_vals = [
                    grid[(r - 1, c - 1)],
                    grid[(r - 1, c)],
                    grid[(r - 1, c + 1)],
                    grid[(r, c - 1)],
                    grid[(r, c + 1)],
                    grid[(r + 1, c - 1)],
                    grid[(r + 1, c)],
                    grid[(r + 1, c + 1)],
                ]

                cnt = neighbor_vals.count("@")
                if cnt < 4:
                    removable.append((r, c))

    return removable


def remove(grid: Grid, positions: list[tuple[int, int]]):
    for pos in positions:
        grid[pos] = "."


def part1(grid: Grid):
    removable = find_removable(grid)

    print("Part 1:", len(removable))


def part2(grid: Grid):
    total_removed = 0

    while True:
        removable = find_removable(grid)
        if len(removable) == 0:
            break

        remove(grid, removable)
        total_removed += len(removable)

    print("Part 2:", total_removed)


filename = "day04/example"
filename = "day04/input"

grid = parse(filename)

part1(grid)
part2(grid)

File: 2025/day04/cpp/day04.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <unordered_map>
#include <filesystem>

using namespace std;


struct PairHash
{
    size_t operator()(const pair<int, int>& p) const noexcept
    {
        // combine two 32-bit ints into a 64-bit value, then hash
        return std::hash<unsigned long long>()(
            (static_cast<unsigned long long>(static_cast<unsigned int>(p.first)) << 32) ^ static_cast<unsigned int>(p.second));
    }
};


class Grid
{
    private:
        unordered_map<pair<int, int>, char, PairHash> map;
		char defaultValue;

    public:
        int RowCnt;
        int ColCnt;

        Grid(char defaultValue)
        {
			this->defaultValue = defaultValue;
        }

        void Set(int row, int col, char value)
        {
            map[{row, col}] = value;
		}

        char GetOrDefault(int row, int col, char defaultValue)
        {
            auto it = map.find({row, col});
            if (it != map.end())
            {
                return it->second;
            }
            return defaultValue;
        }
};


static Grid parse(const string& filename)
{
    string fileInfo;
    fileInfo += "This is the current working directory: " + filesystem::current_path().string() + ".\n";
	fileInfo += "Filename supplied: " + filename + ".\n";
	fileInfo += "Resolves to absolute path: " + filesystem::absolute(filename).string() + ".\n";

    if (!filesystem::exists(filename))
    {
        cerr << fileInfo << endl;
        cerr << "Error: file does not exist: " << filesystem::absolute(filename) << endl;
		throw runtime_error("file does not exist");
    }

    ifstream f(filename);

    Grid grid('@');

    int row = 0;
    int col = 0;

    string line;

    while (getline(f, line))
    {
        col = 0;
        for (char c : line)
        {
            grid.Set(row, col, c);
            col++;
        }
        row++;
    }

	grid.RowCnt = row;
	grid.ColCnt = col;

    return grid;
}



vector<pair<int, int>> find_removable(Grid& grid)
{
    vector<pair<int, int>> removable;

    for (int r = 0; r < grid.RowCnt; r++)
    {
        for (int c = 0; c < grid.ColCnt; c++)
        {
            char curr_val = grid.GetOrDefault(r, c, ' ');
            if (curr_val == '@')
            {
                // get neighbors in all 8 directions
                vector<char> neighbor_vals = {
                    grid.GetOrDefault(r - 1, c - 1, ' '),
                    grid.GetOrDefault(r - 1, c, ' '),
                    grid.GetOrDefault(r - 1, c + 1, ' '),
                    grid.GetOrDefault(r, c - 1, ' '),
                    grid.GetOrDefault(r, c + 1, ' '),
                    grid.GetOrDefault(r + 1, c - 1, ' '),
                    grid.GetOrDefault(r + 1, c, ' '),
                    grid.GetOrDefault(r + 1, c + 1, ' ')
                };
                auto cnt = count(neighbor_vals.begin(), neighbor_vals.end(), '@');
                if (cnt < 4)
                {
                    removable.push_back({r, c});
                }
            }
        }
    }
    return removable;
}


void part1(Grid& grid)
{
    auto removable = find_removable(grid);
    cout << "Part 1:" << removable.size() << endl;
}

void part2(Grid& grid)
{
    size_t total_removed = 0;
    while (true)
    {
        auto removable = find_removable(grid);
        if (removable.size() == 0)
        {
            break;
        }
        for (const auto& pos : removable)
        {
            grid.Set(pos.first, pos.second, ' ');
        }
        total_removed += removable.size();
    }
    cout << "Part 2: " << total_removed << endl;
}

int main()
{
    //string filename = "../../../../example";
    string filename = "../../../../input";

    auto grid = parse(filename);
	part1(grid);
    part2(grid);

	return 0;
}
